server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -115,13 +113,13 @@ private DocumentMapper(IndexSettings indexSettings,
         this.documentParser = documentParser;
         this.indexSettings = indexSettings;
         this.indexAnalyzers = indexAnalyzers;
-        this.fieldMappers = MappingLookup.fromMapping(this.mapping);
 
         try {
             mappingSource = new CompressedXContent(this, XContentType.JSON, ToXContent.EMPTY_PARAMS);
         } catch (Exception e) {
             throw new ElasticsearchGenerationException("failed to serialize source for type [" + type + "]", e);
         }
+        this.fieldMappers = MappingLookup.fromMapping(mapping, this::parse, w -> w.write(mappingSource.compressed()));
```

Comment:
The extra arguments are required so that `MappingLookup` can do all the things we need it to do now.